### DISCUSSION TITLE
cmake fails when cross compiling with mbedtls

### DISCUSSION THREAD

**Discussion by kevinduk** [awslabs/amazon-kinesis-video-streams-webrtc-sdk-c#1919] [Q&A] [answered]
I am trying to cross-compile amazon-kinesis-video-streams-webrtc-sdk-c with mbedTLS but the cmake command fails.

I do not know if I have made a mistake or if it just doesn't work. I am running on Ubuntu 20.04.

I have set CC and CXX for the cross-compiler and want mbedTLS and other dependencies to be built. I have run this command:

cmake -DBUILD_STATIC_LIBS=TRUE -DBUILD_WEBRTC_SAMPLES=ON -DBUILD_LIBSRTP_HOST_PLATFORM=x86_64 -DUSE_OPENSSL=OFF -DUSE_MBEDTLS=ON  ..

The mbedTLS libraries are built correctly.

But when cmake is building libwebsockets there is a compile error:

`/home/kevind/code/amazon-kinesis-video-streams-webrtc-sdk-c-H8XXX/amazon-kinesis-video-streams-webrtc-sdk-c/open-source/libwebsockets/build/src/project_libwebsockets/include/libwebsockets.h:287:25: fatal error: mbedtls/ssl.h: No such file or directory
 #include <mbedtls/ssl.h>
                         ^
compilation terminated.
`

This suggests that the include path is not been passed in the libwebsockets build args.

I also see these messages which shows cmake is picking up mbedtls headers from the Ubuntu installed mbedtls instead of mbedtls built by cmake (which could be a different version).

`Compiling with SSL support
MBEDTLS include dir: /usr/include
MBEDTLS libraries: /home/kevind/code/amazon-kinesis-video-streams-webrtc-sdk-c-H8XXX/amazon-kinesis-video-streams-webrtc-sdk-c/open-source/lib/libmbedtls.a;/home/kevind/code/amazon-kinesis-video-streams-webrtc-sdk-c-H8XXX/amazon-kinesis-video-streams-webrtc-sdk-c/open-source/lib/libmbedx509.a;/home/kevind/code/amazon-kinesis-video-streams-webrtc-sdk-c-H8XXX/amazon-kinesis-video-streams-webrtc-sdk-c/open-source/lib/libmbedcrypto.a`

Does anyone know if my cmake command is incorrect, or should I create an Issue?

**Comment by niyatim23** [comment]
Hi @kevinduk, we have some cross compilation builds in our CI that work fine. Can you refer to these if it helps
https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/bfa6667e2f2eeb800a0edd6e6e4745b4faf34536/.github/workflows/ci.yml#L526-L579

**Reply by kevinduk** [reply]
Hi Niyati, thank you for the pointer. The samples in that file are all for openssl but I am trying to build with mbedtls.

It seems to me that https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/master/CMake/Dependencies/libwebsockets-CMakeLists.txt passes on the openssl include and library paths but not the mbedtls  paths. These lines need adding:

```
        -DLWS_MBEDTLS_INCLUDE_DIRS=${LWS_MBEDTLS_INCLUDE_DIRS}
        -DLWS_MBEDTLS_LIBRARIES=${LWS_MBEDTLS_LIBRARIES}
```
LWS_MBEDTLS_INCLUDE_DIRS and LWS_MBEDTLS_LIBRARIES need to be set to the mbedtls include and library paths in CMakeLists.txt.

I have fixed that on my local checkout and now libwebsockets.a is building. Shall I create an issue?

But now I am getting other linker errors because websockets test app lws-api-test-lejp cannot find the mbedtls libraries, even though LWS_WITHOUT_TESTAPPS=1 . That is my next hurdle.

**Reply by niyatim23** [reply]
Does building libwebsockets with mbedtls standalone work for you?

**Reply by kevinduk** [reply]
If you mean doing a fresh cloning of the libwebsockets repo and building with mbedtls there then I haven't tried. If you can give me the cmake command line for libwebsockets and mbedtls then I can try it.

**Reply by niyatim23** [reply]
Yes, I meant that. I haven't tried it myself but you may be able to find the steps specific to your environment in the [CI](https://github.com/warmcat/libwebsockets/blob/main/.sai.json) for the libwebsockets repository. 

Additionally, our CI does have an example for mbedtls. Can you check if that works for you?
https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/bfa6667e2f2eeb800a0edd6e6e4745b4faf34536/.github/workflows/ci.yml#L325-L391

**Reply by kevinduk** [reply]
> Additionally, our CI does have an example for mbedtls. 

Unfortunately, the examples are not cross compiled. It is the combination of cross-compile and mbedtls that is the problem.

**Reply by niyatim23** [reply]
Understood, can you please share the compiler you're using? Also, are there any other changes except the cross-compilation with mbedtls? I'm going to try to install and build it myself on an Ubuntu instance

**Comment by disa6302** [comment]
@kevinduk ,

Can you confirm what `CC` and `CXX` are being set to? 
In addition to this, can you also confirm which SDK version you are using and if there are any changes to the dependencies version you made locally?

**Comment by kevinduk** [comment, marked as answer]
PR https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/1923 has added cross compile examples.

To cross compile on x86_64 host for arm32 target using mbedtls static libraries:

```
sudo apt-get -y install gcc-arm-linux-gnueabi g++-arm-linux-gnueabi binutils-arm-linux-gnueabi
export CC=arm-linux-gnueabi-gcc
export CXX=arm-linux-gnueabi-gc++
cmake .. -DBUILD_STATIC_LIBS=ON -DUSE_OPENSSL=OFF -DUSE_MBEDTLS=ON -DBUILD_LIBSRTP_HOST_PLATFORM=x86_64-unknown-linux-gnu -DBUILD_LIBSRTP_DESTINATION_PLATFORM=arm-unknown-linux-uclibcgnueabi
make
```

**Reply by mehmetcy84** [reply]
@kevinduk what change has resolved the issue for you? I am also trying to build the Kinesis C SDK for iOS with mbedtls enabled, and the libwebsocket build throws the same error for me:
libwebsockets.h:287:25: fatal error: mbedtls/ssl.h: No such file or directory #include <mbedtls/ssl.h>

When I check the open-source/lib folder, I see that mbedtls is built fine for iOS device (arm64).
